### PR TITLE
P07/feat/token moe interaction analysis

### PR DESCRIPTION
**Pull request by firekind**
## Description

Adds the calculations for the following metrics (for each moe router):
- **`entropy_mean`** (`float`): The mean entropy for this router. Measure of how confident the router was when routing a token. Low entropy → more confident, high entropy → less confident.

- **`entropy_std`** (`float`): Standard deviation of entropy, i.e., its spread.

- **`entropy_per_group`** (`list[float]`): Average entropy for each token group for this router. Value at index *i* is the average entropy of the *i*th group.

- **`n_eff`** (`float`): Effective number of experts being used (out of `num_real_experts` + 1). Calculated as exp(entropy_mean). Includes null expert collapsed as single option. Values close to (`num_real_experts` + 1) indicate uniform routing, close to 1 indicate concentration.

- **`polarization`** (`float`): Measure of bimodality - are tokens getting routed to only real or mostly null experts? Value near 0 = gradual distribution, near 1 = polarized.

- **`tokens_to_real_rate`** (`float`): What fraction of tokens in the batch were sent to real experts? Value between 0.0 and 1.0.

- **`tokens_to_null_rate`** (`float`): What fraction of tokens in the batch were sent to null experts? Value between 0.0 and 1.0.

- **`null_junk_rate`** (`float`): Out of the tokens sent to null expert, what fraction of them were junk? Tells us whether the null experts are staying pure, or if they are also receiving content related tokens. Value will be `NaN`if no tokens were sent to null expert.

- **`junk_to_null_rate`** (`float`): Out of the junk tokens present in the batch, what fraction went to null experts? Tells us whether we are successfully routing junk tokens to null experts. Value will be `NaN` if there are no junk tokens present in the batch.

- **`imbalance_ratio`** (`float`): Ratio between the busiest real expert and the least busy real expert. Lower value → more balanced. Null experts are ignored.

- **`gini`** (`float`): The Gini Coefficient. Measure of inequality in workload distribution between real experts.

- **`cv`** (`float`): Coefficient of variation. Measure of how spread out the workload is among real experts.

- **`starvation_count`** (`int`): Number of real experts receiving less than ε traffic. Expert is starving if it receives tokens < ε * average tokens per expert.

- **`router_logit_scale`** (`float`): Standard deviation of router logits.

Also added a dummy model + train script to test these metrics out. Currently, the metrics are printed to stdout for each step.

> Note: token groups are not finalized. token id to token group mapping is not finalized. Model, dataset and tokenizer used are not representative of the final model, dataset and tokenizer that will be used. They were chosen just to verify calculations.

Also adds a script to calculate token distribution per modality for a dataset.
## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if applicable).
- [x] My code follows the style guidelines, gitflow branching strategy, and naming conventions of this project [Contribution Guidelines](<https://github.com/The-School-of-AI/LLM/tree/main/experiments/>

## Reviewers

- [x] **Reviewer 1**: A member from your own team.
- [x] **Reviewer 2**: A member from the repo owners team (@The-School-of-AI/llm-repo-owners). 

> **Note**: Every pull request requires atleast 2 reviewers/approvers before it can be merged.

### CODE REVIEW

File: uv.lock
Comment (pankajkr23) [inline uv.lock:1]: Please remove uv.lock file

Comment (firekind) [reply uv.lock:1]: done
